examples/substrait-spark/README.md

```diff
@@ -0,0 +1,586 @@
+
+Finally this can be converted to a Spark Plan:
+
+```java
+      ToLogicalPlan substraitConverter = new ToLogicalPlan(spark);
+      LogicalPlan sparkPlan = substraitConverter.convert(plan);
+```
+
+If you were to print out this plan, it has the identical structure to the plan seen earlier on.
+
+```
++- Sort [count(1)#18L ASC NULLS FIRST], true
+   +- Aggregate [colour#5], [colour#5, count(1) AS count(1)#18L]
+      +- Project [colour#5]
+         +- Join Inner, (vehicle_id#2 = vehicle_id#10)
+            :- Project [vehicle_id#2, colour#5]
+            :  +- Filter isnotnull(vehicle_id#2)
+            :     +- Relation [vehicle_id#2,make#3,model#4,colour#5,fuel_type#6,cylinder_capacity#7,first_use_date#8] csv
+            +- Project [vehicle_id#10]
+               +- Filter ((isnotnull(test_result#14) AND (test_result#14 = P)) AND isnotnull(vehicle_id#10))
+                  +- Relation [test_id#9,vehicle_id#10,test_date#11,test_class#12,test_type#13,test_result#14,test_mileage#15,postcode_area#16] csv
+
+```
+
+Executed of this plan is then simple `Dataset.ofRows(spark, sparkPlan).show();` giving the output of
+
+```java
++------+-----+
+|colour|count|
++------+-----+
+| GREEN|    1|
+|BRONZE|    1|
+|   RED|    2|
+| BLACK|    2|
+|  GREY|    2|
+|  BLUE|    2|
+|SILVER|    3|
+| WHITE|    5|
++------+-----+
+```
+
+### Observations
+
+To recap on the steps above
+
+- Two CSV files have been loaded into Spark
+- Using either the Spark SQL or the Spark Dataset API we can produce a query across those two datasets
+- Both queries result in Spark creating a logical and optimized query plan
+  - And both being are structurally identical
+- Using the Substrait-Java library, we can convert the optimized plan into the Substrait format.
+- This Substrait intermediate representation of the query can be serialized via the protobuf format
+  - Here store as a flat file containing the bytes of that protobuf
+- *Separately* this file can be loaded and the Substrait Plan converted to a Spark Plan
+- This can be run in an application on Spark getting the same results
+
+---
+## Plan Comparison
+
+The structure of the query plans for both Spark and Substrait are structurally very similar.
+
+### Aggregate and Sort
+
+Spark's plan has a Project that filters down to the colour, followed by the Aggregation and Sort.
+```
++- Sort [count(1)#18L ASC NULLS FIRST], true
+   +- Aggregate [colour#5], [colour#5, count(1) AS count(1)#18L]
+      +- Project [colour#5]
+```
+
+When converted to Substrait the Sort and Aggregate is in the same order, but there are additional projects; it's not reduced the number of fields as early.
+
+```
++- Sort:: FieldRef#/I64/StructField{offset=1}  ASC_NULLS_FIRST
+   +- Project:: [Str, I64, Str, I64]
+      +- Aggregate:: FieldRef#/Str/StructField{offset=0}
+```
```

Comment:
I'm not sure what is intended to be shown here because these two plans fragments are very different.

Reply:
I've updated the text to explain more the relevance. they have the same origin. The point is to show that (sometimes) Spark has reordered the relations when optimised, and the Substrait project relations inside the plan don't reduce the number of fields, this only happens on the outer relations.

Reply:
Yes - that's completely fine; it's not central to the example;